dvc/utils/strictyaml.py

```diff
@@ -141,49 +150,64 @@ def prepare_code(mark: "StreamMark") -> "Syntax":
-    lines of code for `n` number of steps taken upwards. In a worst case,
-    it may be just 1 step (as non-collection item cannot have child items),
-    but `schema validator` may provide us arbitrary path. So, this caps the
-    number of steps upward to just 5. If it does not find any linecols, it'll
-    abort.
-    """
-    from dpath import get
+def _normalize_linecol(lc: "LineCol | Tuple[Any, Any]") -> Tuple[int, int]:
+    from ruamel.yaml.comments import LineCol
 
-    step = 1
     line, col = None, None
-    while paths and step < max_steps:
-        value = get(data, paths, default=None)
-        if value is not None:
-            with suppress(AttributeError, TypeError):
-                line = value.lc.line + 1  # type: ignore[attr-defined]
-                col = value.lc.col + 1  # type: ignore[attr-defined]
-                break
-        step += 1
-        *paths, _ = paths
 
-    return line, col, step
+    if isinstance(lc, LineCol):
+        line = lc.line
+        col = lc.col
+    elif isinstance(lc, tuple):
+        line = int(lc[0])
+        col = int(lc[1])
+
+    if not (isinstance(line, int) and isinstance(col, int)):
+        raise TypeError(f"Unable to determine neither line nor col ({line}:{col})")
+
+    return line + 1, col + 1
+
+
+def determine_linecol(
+    yaml_data: Any, glob_pattern: Sequence[str], key_or_value: bool = False
+) -> Tuple[int, int]:
+    """
+    Return the line and column number for the given location in the data.
+
+    Args:
+        yaml_data: The data must be a parsed YAML document represented as a Python
+            object, obtained by calling `ruamel.yaml.YAML(typ='rt').load()` on a YAML
+            file. The function expects the data to contain information about the
+            original YAML document's structure, including the line and column numbers
+            where each element appears.
+        glob_pattern: The glob pattern to match a key or item in the data.
+        key_or_value: A boolean flag that controls whether the result points to the key
+            or value of the matching key/value pair. Ignored when matching an item
+            within a sequence. Defaults to False.
+
+    Raises:
+        ValueError: Raised if more than one leaf matches the glob.
+        KeyError: Raised if the location not found.
+
+    Returns:
+        A tuple containing the line and column number of the matched location.
+    """
+    from dpath import get
+    from ruamel.yaml.comments import CommentedMap, CommentedSeq
+
+    obj = get(yaml_data, glob_pattern[:-1])
+    if isinstance(obj, CommentedMap):
+        return _normalize_linecol(
+            obj.lc.key(glob_pattern[-1])
+            if key_or_value
+            else obj.lc.value(glob_pattern[-1])
+        )
+    if isinstance(obj, CommentedSeq):
+        return _normalize_linecol(obj.lc.item(int(glob_pattern[-1])))
+
+    raise TypeError(
+        f"Expected commented seq or map, got {type(obj)} at path {glob_pattern!r}"
+    )
```

Comment:
Note there may be cases where we don't have paths. So this should just return linecol from `data` in that case.